### PR TITLE
ORA1554 Thoroughly test revamp synth pr

### PR DESCRIPTION
**Pull request by br4e**
This is _so close_. All of the test values are doing the thing where they are basically right, except they are about 5% off of their correct values. 

I have noticed that the Forecast implied inferences are a little bit off, and that is throwing off all of the other values, EXCEPT for the naive network inference which is exactly correct, _which does not use forecast implied inferences_

If we can just figure out why forecast implied inferences are not returning the correct value, I think we've got it. 

I've looked at it a few times. Anyone else want to give it a look? You can run `TestForecastImpliedInferences` to test.

### CODE REVIEW

File: x/emissions/keeper/inference_synthesis/network_inference_builder_test.go [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
+	cosmosMath "cosmossdk.io/math"
 
 	// cosmosMath "cosmossdk.io/math"
 	storetypes "cosmossdk.io/store/types"
 	"github.com/allora-network/allora-chain/app/params"
 	alloraMath "github.com/allora-network/allora-chain/math"
+	alloratestutil "github.com/allora-network/allora-chain/test/testutil"
 	"github.com/allora-network/allora-chain/x/emissions/keeper"
-	inferencesynthesis "github.com/allora-network/allora-chain/x/emissions/keeper/inference_synthesis"
+	inference_synthesis "github.com/allora-network/allora-chain/x/emissions/keeper/inference_synthesis"
```
Comment (guilherme-brandao) [inline x/emissions/keeper/inference_synthesis/network_inference_builder_test.go:21]: using it without `_`   or camel case seems to be the right pattern, but nbd

File: x/emissions/keeper/inference_synthesis/network_inference_builder.go [hunk 2]
```diff
@@ -24,16 +24,19 @@ type NetworkInferenceBuilder struct {
...
 func NewNetworkInferenceBuilderFromSynthRequest(
 	req SynthRequest,
-) *NetworkInferenceBuilder {
+) (*NetworkInferenceBuilder, error) {
 	paletteFactory := SynthPaletteFactory{}
-	palette := paletteFactory.BuildPaletteFromRequest(req)
+	palette, err := paletteFactory.BuildPaletteFromRequest(req)
+	if err != nil {
+		return nil, errorsmod.Wrapf(err, "Error building palette from request")
+	}
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/network_inference_builder.go:32]: We should be returning errors

File: x/emissions/keeper/inference_synthesis/network_inferences.go [hunk 3]
```diff
@@ -102,7 +102,11 @@ func GetNetworkInferencesAtBlock(
 				CNorm:               moduleParams.CNorm,
 			},
 		)
-		networkInferences = networkInferecesBuilder.CalcAndSetNetworkInferences().Build()
+		if err != nil {
+			ctx.Logger().Warn(fmt.Sprintf("Error constructing network inferences builder topic: %s", err.Error()))
+			return nil, err
+		}
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/network_inferences.go:108]: We should be returning errors

File: x/emissions/keeper/inference_synthesis/synth_palette_factory.go [hunk 4]
```diff
@@ -27,17 +28,20 @@ func (f *SynthPaletteFactory) BuildPaletteFromRequest(req SynthRequest) SynthPal
...
 	palette.AllWorkersAreNew = palette.AllInferersAreNew && palette.AllForecastersAreNew
 
-	return palette
+	// Populates: forecastImpliedInferenceByWorker,
+	err := palette.UpdateForecastImpliedInferences()
+	if err != nil {
+		return SynthPalette{}, err
+	}
+
+	return palette, nil
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_factory.go:46]: There were a lot of bugs caused by these operations being out of order, also we should return errors

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 5]
```diff
@@ -15,8 +15,8 @@ func (p *SynthPalette) CalcWeightsGivenWorkers() (RegretInformedWeights, error)
 	if err != nil {
 		return RegretInformedWeights{}, errorsmod.Wrapf(err, "Error calculating standard deviation of regrets")
 	}
-	// Add epsilon to standard deviation
-	stdDevRegretsPlusEpsilon, err := stdDevRegrets.Abs().Add(p.Epsilon)
+	// Add f Tolerance to standard deviation
+	stdDevRegretsPlusEpsilon, err := stdDevRegrets.Abs().Add(p.FTolerance)
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:19]: Use fTolerance in this equation as Guilherme suggested

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 6]
```diff
@@ -111,7 +111,7 @@ func (p *SynthPalette) CalcWeightedInference(weights RegretInformedWeights) (Inf
 
 	for _, forecaster := range p.Forecasters {
 		runningUnnormalizedI_i, sumWeights, err = AccumulateWeights(
-			p.InferenceByWorker[forecaster],
+			p.ForecastImpliedInferenceByWorker[forecaster],
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:114]: This is a big one!

**Review by br4e — COMMENTED** [review]
noting important changes on this pr

File: x/emissions/keeper/inference_synthesis/synth_palette_factory.go (hunk 4 above)
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_factory.go:46]: There were a lot of bugs caused by these operations being out of order, also we should return errors

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 7]
```diff
@@ -111,7 +111,7 @@ func (p *SynthPalette) CalcWeightedInference(weights RegretInformedWeights) (Inf
...
 			weights.forecasters[forecaster],
 			p.ForecasterRegrets[forecaster].noPriorRegret,
 			p.AllForecastersAreNew,
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:117]: This is a big one!

File: x/emissions/keeper/inference_synthesis/synth_palette_weight.go [hunk 8]
```diff
@@ -36,13 +36,13 @@ func (p *SynthPalette) CalcWeightsGivenWorkers() (RegretInformedWeights, error)
...
 		regretFrac, err := p.ForecasterRegrets[worker].regret.Quo(stdDevRegretsPlusEpsilon)
 		if err != nil {
 			return RegretInformedWeights{}, errorsmod.Wrapf(err, "Error calculating regret fraction")
 		}
 		normalizedForecasterRegrets[worker] = regretFrac
-		if i == 0 || regretFrac.Gt(maxRegret) {
+		if regretFrac.Gt(maxRegret) {
 			maxRegret = regretFrac
 		}
 	}
```
Comment (br4e) [inline x/emissions/keeper/inference_synthesis/synth_palette_weight.go:48]: We don't want to re-initialize the max value

**Review by kpeluso — APPROVED** [review]

**Pull request merged**